be/src/io/direct_s3_output_stream.cpp

```diff
@@ -0,0 +1,123 @@
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/CompleteMultipartUploadRequest.h>
+#include <aws/s3/model/CreateMultipartUploadRequest.h>
+#include <aws/s3/model/PutObjectRequest.h>
+#include <aws/s3/model/UploadPartRequest.h>
+#include <fmt/format.h>
+
+#include "common/logging.h"
+
+namespace starrocks::io {
+
+class StringViewStream : Aws::Utils::Stream::PreallocatedStreamBuf, public std::iostream {
+public:
+    StringViewStream(const void* data, int64_t nbytes)
+            : Aws::Utils::Stream::PreallocatedStreamBuf(reinterpret_cast<unsigned char*>(const_cast<void*>(data)),
+                                                        static_cast<size_t>(nbytes)),
+              std::iostream(this) {}
+};
+
+DirectS3OutputStream::DirectS3OutputStream(std::shared_ptr<Aws::S3::S3Client> client, std::string bucket,
+                                           std::string object)
+        : _client(std::move(client)), _bucket(std::move(bucket)), _object(std::move(object)) {
+    DCHECK(_client != nullptr);
+}
+
+Status DirectS3OutputStream::write(const void* data, int64_t size) {
+    if (_upload_id.empty()) {
+        RETURN_IF_ERROR(create_multipart_upload());
+        DCHECK(!_upload_id.empty());
+    }
+
+    if (size == 0) {
+        return Status::OK();
+    }
+
+    Aws::S3::Model::UploadPartRequest req;
+    req.SetBucket(_bucket);
+    req.SetKey(_object);
+    req.SetPartNumber(static_cast<int>(_etags.size() + 1));
+    req.SetUploadId(_upload_id);
+    req.SetContentLength(size);
+    req.SetBody(std::make_shared<StringViewStream>(data, size));
+    auto outcome = _client->UploadPart(req);
+    if (!outcome.IsSuccess()) {
+        return Status::IOError(
+                fmt::format("S3: Fail to upload part of {}/{}: {}", _bucket, _object, outcome.GetError().GetMessage()));
+    }
+
+    _etags.push_back(outcome.GetResult().GetETag());
+    return Status::OK();
+}
+
+Status DirectS3OutputStream::close() {
+    if (_client == nullptr) {
+        return Status::OK();
+    }
+
+    if (!_upload_id.empty()) {
+        RETURN_IF_ERROR(complete_multipart_upload());
+    }
+
+    _client = nullptr;
+    return Status::OK();
+}
+
+Status DirectS3OutputStream::create_multipart_upload() {
+    Aws::S3::Model::CreateMultipartUploadRequest req;
+    req.SetBucket(_bucket);
+    req.SetKey(_object);
+    Aws::S3::Model::CreateMultipartUploadOutcome outcome = _client->CreateMultipartUpload(req);
+    if (outcome.IsSuccess()) {
+        _upload_id = outcome.GetResult().GetUploadId();
+        return Status::OK();
+    }
+    return Status::IOError(fmt::format("S3: Fail to create multipart upload for object {}/{}: {}", _bucket, _object,
+                                       outcome.GetError().GetMessage()));
+}
+
+Status DirectS3OutputStream::complete_multipart_upload() {
+    VLOG(12) << "Completing multipart upload s3://" << _bucket << "/" << _object;
+    DCHECK(!_upload_id.empty());
+    DCHECK(!_etags.empty());
```

Comment:
what will happen when we create multiupload successfully, but uploadpart failed which lead to no etags(all failed) or only partial etags(some failed), will we still close or shall we clear?

Reply:
I primarily reference the implementation of `be/src/io/direct_s3_output_stream.cpp`. It should work.